src/pymor/algorithms/timestepping.py

```diff
@@ -10,221 +10,602 @@
+            The interpolation point within the latest computed step interval.
+
+        Returns
+        -------
+        U
+            A |VectorArray| of length 1 containing U(t).
+        """
+        t_n, t_np1 = self.t_n, self.t_np1
+        assert floatcmp.almost_less(t_n, t)
+        assert floatcmp.almost_less(t, t_np1)
+        if self.stepper.interpolation == 'P0':
+            # return previous value, old default in pyMOR
+            return self.U_n.copy()
+        elif self.stepper.interpolation == 'P1':
+            # compute P1-Lagrange interpolation
+            U_n, U_np1 = self.U_n, self.U_np1
+            dt = t_np1 - t_n
+            return (t_np1 - t)/dt * U_n + (t - t_n)/dt * U_np1
+        else:
+            raise NotImplementedError(f'{self.interpolation}-interpolation not available!')
+
+    def _interpolated_step(self):
+        """
+        Returns `(U_next, t_next)` (if `return_times == True`, otherwise `U_next`), behaviour depends on `num_values`:
+        - If `num_values` is provided, performs as many actual steps (by calling :meth:`_step`) as required to obtain
+          an interpolation of the solution, U(t_next) at the next required interpolation point t_next >= t_n, and
+          returns (U(t_next), t_next).
+        - If `num_values` is not provided, performs a single step (by calling :meth:`_step`) starting from t_n, to
+          compute (U(t_{n + 1}), t_{n + 1}).
+        """
+        if self.t > self.t1 or floatcmp.float_cmp(self.t, self.t1):
+            # this is the end
+            raise StopIteration
+        elif self.stepper.num_values:
+            # an interpolation of the trajectory is requested
+            if self._last_stepped_point < self.t0:
+                # this is the start, take a step to have data and interpolation available next time, but return U0
+                _, self.t = self._step()
+                self._last_stepped_point = self.t
+                self._next_interpolation_point += self._save_dt
+                if self.return_times:
+                    return self.U0, self.t0
+                else:
+                    return self.U0
+            elif self._last_stepped_point > self._next_interpolation_point:
+                # we have enough data, simply interpolate
+                if not self.logging_disabled:
+                    self.logger.debug(f't={self._next_interpolation_point}: interpolating ...')
+                t_next = self._next_interpolation_point
+                U_next = self._interpolate(t_next)
+                self._next_interpolation_point += self._save_dt
+                if self.return_times:
+                    return U_next, t_next
+                else:
+                    return U_next
+            else:
+                # we do not have enough data, take enough actual steps
+                while self.t < self._next_interpolation_point:
+                    _, self.t = self._step()
+                    self._last_stepped_point = self.t
```

Comment:
It might happen that your last stepped point is `t_np1`, but due to `num_values`, some intermediate values `t < t_np1` are required by interpolation. In that case, in each iteration `t` would correspond to the value of the actually returned `U`, while `_last_stepped_point` would remain at `t_np1` until we are there.